### PR TITLE
Scroll on drag

### PR DESCRIPTION
**Pull request by Klaasvaak**
From issue: https://github.com/bevacqua/dragula/issues/121#issuecomment-138992160

## TODO:
- [x] vertical scroll
- [ ] horizontal scroll
- [ ] possibility to disable scroll
- [ ] possibility to add the scroll triggers
- [ ] possibility to override the scrollfunction
- [ ] check touch
- [ ] ~~check if we should scroll on gu-mirror position instead of mouse position?~~ (could break when containers are too close to each other)

### CODE REVIEW

**Comment by Klaasvaak** [comment]
I added an example at the bottom of the example page.

File: dragula.js [hunk 1]
```diff
@@ -62,6 +66,11 @@ function dragula (initialContainers, options) {
     var op = remove ? 'remove' : 'add';
     touchy(documentElement, op, 'mousedown', grab);
     touchy(documentElement, op, 'mouseup', release);
+    if (remove) {
+      document.removeEventListener('mousemove', onMouseMove);
+    } else {
+      document.addEventListener('mousemove', onMouseMove);
```
Comment (bevacqua) [inline dragula.js]: You should replace all of this for `touchy(document, op, 'mousemove', onMouseMove);`

File: dragula.js [hunk 2]
```diff
@@ -445,6 +454,56 @@ function dragula (initialContainers, options) {
...
+        } else if (y >= bottomTriggerContainer.top && y <= bottomTriggerContainer.bottom
+          && x >= bottomTriggerContainer.left && x <= bottomTriggerContainer.right) {
+          scrollDown(container, y);
+        }
+      }
+    }
+  }
+
+  function scrollDown (container, y) {
+    if (y === drake.previousY) {
```
Comment (bevacqua) [inline dragula.js]: Let's create local variables inside the `dragula` method instead of exposing them on `drake`.

File: dragula.js (hunk 1 above)
Comment (Klaasvaak) [inline dragula.js]: Yes, makes sense. Will do

File: dragula.js (hunk 2 above)
Comment (Klaasvaak) [inline dragula.js]: Will do.

**Comment by btecu** [comment]
@Klaasvaak What's the state of this?

**Comment by Klaasvaak** [comment]
@btecu I didn't forget it. Not sure when I got time. Well give it an update soon though.

**Comment by bevacqua** [comment]
Could someone pick this one up?

**Comment by Klaasvaak** [comment]
If someone has got more time I can give them access to the fork, so they can help with this PR.

**Comment by hollowdoor** [comment]
I think this should be a separate module. See how gtk does it: [gtk scroll window](http://www.gtk.org/tutorial1.2/gtk_tut-10.html#ss10.9). It would be more involved as a javascript module with all the possible events, but I think it's doable.

Then dragula can require the "scrollable", or users can make a dragula container scrollable from their own code.

A generic scrollable element like this could be used elsewhere as well. It would also be easier to maintain in the long run.

Also there are already some modules that do this. For example: [zenscroll](https://github.com/zengabor/zenscroll) which can scroll to an element, or pixel position. There are others, but most of those are plugins for other frameworks. It's a lot to search through.

**Comment by bevacqua** [comment]
+1 keeping it out of dragula. Would be nice to have one to point people to, tho

**Comment by hollowdoor** [comment]
I agree. There are modules. [More than a hundred](https://www.npmjs.com/search?q=scrollable&page=1) in fact that do scrolling found in npm. I'm searching through them now. If I don't find one that works well with dragula I might have to make it myself unless someone can get to it before me. Right now I'm considering zenscroll.

**Comment by Klaasvaak** [comment]
@hollowdoor If it's possible to keep it out than I agree. I needed this for a project and it wasn't hard to do. But still it would be nice if it worked out of the box.

**Comment by hollowdoor** [comment]
I found this thread because of the same thing. I was working with dragula, and it wouldn't scroll my div. That's a problem for dragula I thought. But after some more thinking it breaks reusability, and least responsibility because scrolling elements is such a common problem, and dragula is not **scrollula**.

Like I said dragula could still require the module, but at least it's algorithms probably shouldn't be in dragula. I don't want to be dogmatic. So any good arguments to the contrary are welcome.

In the mean time I'm searching for a third party module that will do exactly what we're talking about. I'm trying [zenscroll](https://github.com/zengabor/zenscroll), and [iscroll](https://github.com/cubiq/iscroll) starting today.

**Comment by btecu** [comment]
Since probably most people using `dragula` would like their containers to be scrollable - because it makes sense from a UX perspective, I think it should be included with `dragula`.

Also, one of the nice ideas of `dragula` is that is simple and small... Pulling in another library may add more to the size than just having something custom.

Either way, having something, even if it depends on an external library is better than nothing (right now).

**Comment by hollowdoor** [comment]
@btecu Right now there are not that many libs that support scrolling without a framework. Maybe the time is ripe for a module that does this as simple as dragula. You know something that handles the most common usages like scrolling while dragging, but stays small.

If dragula gets the code that would mean problems would bubble up from scrolling to where the two functionalities of scrolling, and dragging are coupled. The complexity has the potential to be horrendous with an increased potential of feature creep. Then we'd be waiting even longer for feature requests, and bug fixes.

The other side of the coin is a dependency could break without warning, but I'm thinking it would be best not to require a scroller at the module level anyway. Simple interoperability is possible.

These are the most common problems in scrolling:

* events (There's a whole lot.)
    * mouse
    * touch
    * dragging in both of those.
    * multi-environment support
* scroll to show element
* scroll to pixel
* animations
    * speed
    * acceleration
* css conflicts (where pesky positions do not like eachother)

I have a feeling that everybody that wants to have scrolling in dragula are thinking of their own problem which is slightly different from everybody else's. This would be where the feature creep would come in, and down the line we'd be wishing there were an independent scrolling module.

The issue appears to be pervasive, but that's why it should be its own module. I can't predict the future. I could be wrong, but I don't think so.

Right now the main usage of scrolling would be getting the next/last out of scroll element into view. I see dragons in those waters.

**Comment by btecu** [comment]
@hollowdoor http://rubaxa.github.io/Sortable/ actually does scroll as well and it does it very well.

I don't agree. When you have more elements in your list than you can possible see and you want to move one at the bottom, it's a pain currently. I think this is what most people are after.

**Comment by hollowdoor** [comment]
@btecu Most people is a present tense assumption. It's up to bevacqua if he wants to maintain that particular functionality as an internal. I see modularity as the way to go, and others will disagree. I've said my piece.

Meanwhile I'm going to see about modulization of scrollables for long term maintainability.

**Comment by Klaasvaak** [comment]
@hallowdoor how would you like to see this in code for people to use . Like this?:
```
var dragula = require('dragula');
var scrollableDragula = require('dragula-scroll');

var drake = dragula({ containers: containers });
var scrollableDrake = scrollableDragula(drake);
```

**Comment by hollowdoor** [comment]
@Klaasvaak That would be great. Right now I'm working on something more generic, but your idea would be very useful for when dragula just needs to scroll without contributing to bloat. Perfect solution IMHO.

Would it be too dark to call it casket? :)

**Comment by bevacqua** [comment]
Closing in favor of modules such as [`dom-autoscroller`](https://www.npmjs.com/package/dom-autoscroller)

**Comment by Klaasvaak** [comment]
:+1:

**Pull request closed without merging**